src/services/telemetry/ConversationTelemetryService.ts

```diff
@@ -0,0 +1,291 @@
+ */
+export class ConversationTelemetryService {
+	private static instance: ConversationTelemetryService
+	private enabled: boolean = false
+	private distinctId: string
+	private clineApiKey?: string
+	private apiEndpoint: string = "https://api.cline.bot/v1/traces"
+	private tracerProvider: NodeTracerProvider | undefined
+	private tracer: any
+	private messageIndices: Map<string, number> = new Map()
+
+	private constructor(distinctId: string) {
+		this.distinctId = distinctId
+	}
+
+	public static getInstance(distinctId: string): ConversationTelemetryService {
+		if (!ConversationTelemetryService.instance) {
+			ConversationTelemetryService.instance = new ConversationTelemetryService(distinctId)
+		}
+		return ConversationTelemetryService.instance
+	}
+
+	public updateTelemetryState(enabled: boolean, clineApiKey?: string): void {
+		console.log("[ConversationTelemetry] Updating telemetry state...", { enabled, clineApiKey })
+
+		// First update the state variables
+		this.enabled = enabled
+		this.clineApiKey = clineApiKey
+
+		// Then initialize the tracer if needed
+		if (this.enabled && !this.tracer) {
+			this.initializeTracer()
+		}
+	}
+
+	private initializeTracer(): void {
+		try {
+			// Create a resource that identifies our service
+			const resource = new Resource({
+				[ATTR_SERVICE_NAME]: "cline-extension",
+				[ATTR_SERVICE_VERSION]: "1.0.0",
+			})
+
+			console.log("[ConversationTelemetry] Initializing OpenTelemetry tracer...", { "this.clineApiKey": this.clineApiKey })
+
+			// Configure the OTLP exporter
+			const headers: Record<string, string> = {
+				"Content-Type": "application/json",
+			}
+
+			// Add API key to headers if available
+			if (this.clineApiKey) {
+				headers["Authorization"] = `Bearer ${this.clineApiKey}`
+			}
+
+			const exporter = new OTLPTraceExporter({
+				url: this.apiEndpoint,
+				headers,
+			})
+
+			// Create the span processor
+			const spanProcessor = new SimpleSpanProcessor(exporter as any)
+
+			// Create the trace provider with the span processor in the config
+			this.tracerProvider = new NodeTracerProvider({
+				resource,
+				spanProcessors: [spanProcessor as any],
+			})
+
+			// Register the provider
+			this.tracerProvider.register()
+
+			// Get a tracer
+			this.tracer = trace.getTracer("cline-conversation-tracer")
+
+			console.log("[ConversationTelemetry] OpenTelemetry tracer initialized successfully")
+		} catch (error) {
+			console.error("[ConversationTelemetry] Failed to initialize OpenTelemetry tracer:", error)
+		}
+	}
+
+	/**
+	 * Captures a message in the conversation as an OpenTelemetry span
+	 * ONLY HAPPENS IF USER IS OPTED INTO CONVERSATION TELEMETRY IN ADVANCED SETTINGS
+	 */
+	public captureMessage(taskId: string, message: TelemetryChatMessage, metadata: ConversationMetadata): void {
+		// Do NOT capture message if user has not explicitly opted in
+		if (!this.enabled || !this.tracer) {
+			return
+		}
```

Comment:
I would think we would want to check this before calling this method.  I know that this consolidates the check to a single place, but it also means that we add one more step for the runtime to process.  It would be interesting to see if the performance hit is worth the maintenance cost or not.